processor/metricstarttimeprocessor/internal/datapointstorage/timeseries_map.go

```diff
@@ -0,0 +1,103 @@
+import (
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+// TimeseriesInfo contains the information necessary to adjust from the initial point and to detect resets.
+type TimeseriesInfo struct {
+	Mark bool
+
+	Number               pmetric.NumberDataPoint
+	Histogram            pmetric.HistogramDataPoint
+	ExponentialHistogram pmetric.ExponentialHistogramDataPoint
+	Summary              pmetric.SummaryDataPoint
+}
+
+type TimeseriesKey struct {
+	Name           string
+	Attributes     [16]byte
+	AggTemporality pmetric.AggregationTemporality
+}
+
+// TimeseriesMap maps from a timeseries instance (metric * label values) to the timeseries info for
+// the instance.
+type TimeseriesMap struct {
+	sync.RWMutex
+	// The mutex is used to protect access to the member fields. It is acquired for the entirety of
+	// AdjustMetricSlice() and also acquired by gc().
+
+	Mark   bool
+	TsiMap map[TimeseriesKey]*TimeseriesInfo
+}
+
+// Get the TimeseriesInfo for the timeseries associated with the metric and label values.
+func (tsm *TimeseriesMap) Get(metric pmetric.Metric, kv pcommon.Map) (*TimeseriesInfo, bool) {
+	// This should only be invoked be functions called (directly or indirectly) by AdjustMetricSlice().
+	// The lock protecting tsm.tsiMap is acquired there.
+	name := metric.Name()
+	key := TimeseriesKey{
+		Name:       name,
+		Attributes: getAttributesSignature(kv),
+	}
+	switch metric.Type() {
+	case pmetric.MetricTypeHistogram:
+		// There are 2 types of Histograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.Histogram().AggregationTemporality()
+	case pmetric.MetricTypeExponentialHistogram:
+		// There are 2 types of ExponentialHistograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		key.AggTemporality = metric.ExponentialHistogram().AggregationTemporality()
+	}
+
+	tsm.Mark = true
+	tsi, ok := tsm.TsiMap[key]
+	if !ok {
+		tsi = &TimeseriesInfo{}
+		tsm.TsiMap[key] = tsi
+	}
+	tsi.Mark = true
+	return tsi, ok
+}
+
+// Create a unique string signature for attributes values sorted by attribute keys.
+func getAttributesSignature(m pcommon.Map) [16]byte {
+	clearedMap := pcommon.NewMap()
+	m.Range(func(k string, attrValue pcommon.Value) bool {
+		value := attrValue.Str()
+		if value != "" {
+			clearedMap.PutStr(k, value)
+		}
+		return true
+	})
+	return pdatautil.MapHash(clearedMap)
+}
+
+// Remove timeseries that have aged out.
+func (tsm *TimeseriesMap) GC() {
+	tsm.Lock()
+	defer tsm.Unlock()
+	for ts, tsi := range tsm.TsiMap {
+		if !tsi.Mark {
+			delete(tsm.TsiMap, ts)
+		} else {
+			tsi.Mark = false
+		}
+	}
+	tsm.Mark = false
+}
+
+func newTimeseriesMap() *TimeseriesMap {
+	return &TimeseriesMap{Mark: true, TsiMap: map[TimeseriesKey]*TimeseriesInfo{}}
```

Comment:
Maybe doesn't have to be done in this PR, but might be good to come up with a block-allocation strategy for `TsiMap` so that the runtime doesn't have to do too much extra work resizing the map on new timeseries being added. Microbenchmarks would probably need to be added before you could effectively measure how useful that would be though.

Reply:
Thats a good idea, but sounds like a bit pre-mature optimization in this PR. Also not sure its necessary yet but I'll create an issue regardless

Reply:
It's fine not to address until it's measurable at least. Might be a premature optimization, I just like to keep an eye out for obvious allocation paths like this because they can often be missed as opportunities down the line.

Reply:
filed this for now for tracking this: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38623